lib/opam/opam_make_opam.ml

```diff
@@ -0,0 +1,162 @@
+               (List.fold ~init:""
+                  ~f:(fun acc el -> acc ^ "    [" ^ el ^ "]\n") cmds) ^
+               "]\n"
+
+let create_build_cmds
+  : String.t List.t -> String.t =
+  function
+  | [] -> format_cmds ["make build"]
+  | build_cmds -> format_cmds build_cmds
+
+let create_install_cmds
+  : String.t List.t -> String.t =
+  function
+  | [] -> format_cmds ["make install"]
+  | install_cmds -> format_cmds install_cmds
+
+let create_remove_cmds
+  : String.t List.t -> String.t =
+  function
+  | [] -> format_cmds ["make remove"]
+  | remove_cmds -> format_cmds remove_cmds
+
+let format_build_dependencies
+  : String.t List.t -> String.t =
+  List.fold ~init:""
+    ~f:(fun acc el -> acc ^ " \"" ^ el ^ "\" {build}")
+
+let format_dependencies
+  : String.t List.t -> String.t =
+  List.fold ~init:""
+    ~f:(fun acc el -> acc ^ " \"" ^ el ^ "\"")
+
+let write_opam
+  : target_dir:String.t -> name:String.t -> semver:String.t
+  -> license:String.t -> maintainer:String.t Option.t -> author:String.t Option.t
+  -> homepage:String.t Option.t -> bug_reports:String.t Option.t -> dev_repo:String.t
+  -> build_cmds:String.t List.t -> install_cmds:String.t List.t
+  -> remove_cmds:String.t List.t -> depends:String.t List.t
+  -> build_depends:String.t List.t -> (Unit.t, Exn.t) Deferred.Result.t =
+  fun ~target_dir ~name ~semver ~license ~maintainer
+    ~author ~homepage ~bug_reports ~dev_repo ~build_cmds ~install_cmds
+    ~remove_cmds ~depends ~build_depends ->
+    let maintainer_str = match maintainer with
+      | Some m -> "maintainer: \"" ^ m ^ "\"\n"
+      | None -> "" in
+    let author_str = match author with
+      | Some a -> "author: \"" ^ a ^ "\"\n"
+      | None -> "" in
+    let homepage_str = match homepage with
+      | Some h -> "homepage: \"" ^ h ^ "\"\n"
+      | None -> "" in
+    let bug_str = match bug_reports with
+      | Some b -> "bug-reports: \"" ^ b ^ "\"\n"
+      | None -> "" in
+    let contents = "opam-version: \"1.2\"\n" ^
+                   "name: \"" ^ name ^ "\"\n" ^
+                   "version: \"" ^ semver ^ "\"\n" ^
+                   maintainer_str ^
+                   author_str ^
+                   homepage_str ^
+                   bug_str ^
+                   "license: \"" ^ license ^ "\"\n" ^
+                   "dev-repo: \"" ^ dev_repo ^ "\"\n" ^
+                   "\n" ^
+                   "build: " ^ (create_build_cmds build_cmds) ^ "\n" ^
+                   "install: " ^ (create_install_cmds install_cmds) ^ "\n" ^
+                   "remove: " ^ (create_remove_cmds remove_cmds) ^ "\n" ^
+                   "\n" ^
+                   "depends: [" ^ ((format_build_dependencies build_depends) ^
+                                   (format_dependencies depends)) ^ "]\n" in
+    Common.Files.write target_dir "opam" contents
+
+let get_target_dir
+  : target_dir: String.t Option.t -> root_file:String.t ->
+  (String.t, Exn.t) Deferred.Result.t =
+  fun ~target_dir ~root_file ->
```

Comment:
The only thing I don't like about that is that the project root search has the potential to be somewhat expensive and Option.value isn't lazy :/ it may not be a big blocker